source/opt/add_var_for_funccall_param.cpp

```diff
@@ -0,0 +1,74 @@
+// Copyright (c) 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "source/opt/add_var_for_funccall_param.h"
+
+#include "ir_builder.h"
+
+using namespace spvtools;
+using namespace opt;
+
+Pass::Status AddVarsForFuncCallParamPass::Process() {
+  auto result = Status::SuccessWithoutChange;
+  auto funcsNum = get_module()->end() - get_module()->begin();
+  if (funcsNum == 1) return result;
+  for (auto& func : *get_module()) {
+    // Get Variable insertion point
+    Instruction* varInsertPt = &*(func.begin()->begin());
+    func.ForEachInst([this, &result, &varInsertPt](Instruction* inst) {
+      if (inst->opcode() == SpvOpFunctionCall) {
+        Instruction* nextInst = inst->NextNode();
+        for (uint32_t i = 0; i < inst->NumInOperands(); ++i) {
+          auto& op = inst->GetInOperand(i);
+          if (op.type != SPV_OPERAND_TYPE_ID) continue;
+          Instruction* op_inst = get_def_use_mgr()->GetDef(op.AsId());
+          if (op_inst->opcode() == SpvOpAccessChain) {
+            InstructionBuilder builder(
+                context(), inst,
+                IRContext::kAnalysisDefUse |
+                    IRContext::kAnalysisInstrToBlockMapping);
+
+            Instruction* op_ptr_type =
+                get_def_use_mgr()->GetDef(op_inst->type_id());
+            Instruction* op_type = get_def_use_mgr()->GetDef(
+                op_ptr_type->GetSingleWordInOperand(1));
+            auto varType = context()->get_type_mgr()->FindPointerToType(
+                op_type->result_id(), SpvStorageClassFunction);
```

Comment:
This part is too long in a single loop ..

I guess you want to replace:
```
// Before:
.. OpFunctionCall .. %access_chain

// After
%tmp = OpVariable .. Function
%load = OpLoad .. %access_chain
OpStore %tmp %load
.. OpFunctionCall .. %tmp
%load2 = OpLoad .. %tmp
OpStore %access_chain %load2
```

If you separated this part in a separate method with a clear name and comment to describe what it does, it would be easier to understand.